Documentation/design-docs/object-stack-allocation.md

```diff
@@ -0,0 +1,190 @@
+
+When a field of an object is modified the jit may need to issue a write barrier:
+* The reference always points to a heap object: normal write barrier should be used
+* The reference always points to a stack object: no write barrier is needed
+* The reference may point to a heap object or to a stack object depending on control flow: checked write barrier should be used
+
+## Existing Prototypes
+
+@echesakovMSFT implemented a [prototype](https://github.com/echesakovMSFT/coreclr/tree/StackAllocation) in 2016.
+
+The goal of the prototype was to have the optimization working end-to-end with a number of simplifications:
+* A simple intra-procedural escape analysis based on
+[[1]](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.73.4799&rep=rep1&type=pdf)
+but without field edges in the connection graph.
+* All call arguments are assumed to be escaping.
+* Only simple objects are stack allocated, arrays of constant size are not analyzed.
+* Only objects that are allocated unconditionally in the method are moved to the stack. An improvement here would
+be allocating other objects dynamically on the stack.
+* If at least one object in a method is stack allocated, all objects are conservatively reported as as TYPE_GC_BYREF
+and a checked write barrier is used in the method.
+* All objects allocated on the stack also have a pre-header allocated. Pre-header is used for synchronization
+and hashing so we could eliminate it if we proved the object wasn't used for synchronization and hashing.
+
+We ran the prototype on System.Private.CoreLib via crossgen in 2016 and 
+objects at 21 allocation sites were moved to the stack.
+
+We also ran an experiment where we changed the algorithm to optimistically assume that no arguments escape.
+The goal was to get an upper bound on the number of potential stack allocations.
+
+* 424 methods out of 6150 had allocations moved to the stack.
+* 586 allocation sites out of 7977 were moved to the stack.
+* One finding was that most sites were exception object allocations (5345), which almost never happen dynamically.
+* Excluding exception allocations we had 586 allocation sites out of 2632 that could be moved to the stack.
+So the upper bound from this experiment is 22.2%.
+
+@AndyAyersMS recently resurrected @echesakovMSFT work and used it to [prototype stack allocation of a simple delegate that's
+directly invoked](https://github.com/dotnet/coreclr/compare/master...AndyAyersMS:NonNullPlusStackAlloc). It exposed a number of things that need to be
+done in the jit to generate better code for stack-allocated objects. The details are in comments of
+[coreclr #1784](https://github.com/dotnet/coreclr/issues/1784).
+
+We did some analysis of Roslyn csc self-build to see where this optimization may be beneficial. One hot place was found in [GreenNode.WriteTo](https://github.com/dotnet/roslyn/blob/fab7134296816fc80019c60b0f5bef7400cf23ea/src/Compilers/Core/Portable/Syntax/GreenNode.cs#L647).
```

Comment:
How hard is to fix this up to save this allocation? It looks pretty straightforward.

Reply:
Yes, implementing a simple struct version of a stack and using it there will remove these allocations.

Reply:
It would be interesting to look at which cases out of the ones you have listed are hard or impossible to fix by a simple local change. I think they would be the ones to focus on. Maybe the delegates?

Reply:
Delegates is one case. Another common case is fixed-length array passed to, e.g., Console.WriteLine.

Reply:
> Another common case is fixed-length array passed to, e.g., Console.WriteLine.

There's also a proposal to avoid that array allocation from the language side in https://github.com/dotnet/csharplang/issues/1757 by allowing `params Span<T>` parameters.

Reply:
Just so i understand, in the WriteTo case, the proposed optimization would remove the allocation of the Stack object itself, but the underlying array it uses internally would still be heap allocated right?  There's no concept of the "entire stack" structure somehow being able to fit on the real execution stack as long as possible, right?

Reply:
Yes, the proposed optimization would remove the allocation of the Stack object itself. "Inlining" object fields into enclosing objects is beyond the scope of this.